x/concentrated-liquidity/internal/math/tick.go

```diff
@@ -1,52 +1,173 @@
+		// Otherwise, from tick 0 to tick -(geometricExponentIncrementDistanceInTicks), we would use the same exponent as the exponentAtPriceOne
+		exponentAtCurrentTick = exponentAtCurrentTick.Sub(sdk.OneInt())
+	}
 
-	return sqrtPrice, nil
+	// Knowing what our exponentAtCurrentTick is, we can then figure out what power of 10 this exponent corresponds to
+	// We need to utilize bigDec here since increments can go beyond the 10^-18 limits set by the sdk
+	currentAdditiveIncrementInTicks := powTenBigDec(exponentAtCurrentTick)
+
+	// Now, starting at the minimum tick of the current increment, we calculate how many ticks in the current geometricExponent we have passed
+	numAdditiveTicks := tickIndex.ToDec().Sub(geometricExponentDelta.ToDec().Mul(geometricExponentIncrementDistanceInTicks))
+
+	// Finally, we can calculate the price
+	price = powTen(geometricExponentDelta).Add(osmomath.BigDecFromSDKDec(numAdditiveTicks).Mul(currentAdditiveIncrementInTicks).SDKDec())
+
+	return price, nil
 }
 
 // PriceToTick takes a price and returns the corresponding tick index
-func PriceToTick(price sdk.Dec) sdk.Int {
-	tick := osmomath.BigDecFromSDKDec(price).CustomBaseLog(TickBase)
-	return tick.SDKDec().TruncateInt()
+func PriceToTick(price sdk.Dec, exponentAtPriceOne sdk.Int) (sdk.Int, error) {
+	if price.Equal(sdk.OneDec()) {
+		return sdk.ZeroInt(), nil
+	}
+
+	if price.IsNegative() {
+		return sdk.Int{}, fmt.Errorf("price must be greater than zero")
+	}
+
+	if exponentAtPriceOne.LT(types.ExponentAtPriceOneMin) || exponentAtPriceOne.GT(types.ExponentAtPriceOneMax) {
+		return sdk.Int{}, types.ExponentAtPriceOneError{ProvidedExponentAtPriceOne: exponentAtPriceOne, PrecisionValueAtPriceOneMin: types.ExponentAtPriceOneMin, PrecisionValueAtPriceOneMax: types.ExponentAtPriceOneMax}
+	}
+
+	// The formula is as follows: geometricExponentIncrementDistanceInTicks = 9 * 10**(-exponentAtPriceOne)
+	// Due to sdk.Power restrictions, if the resulting power is negative, we take 9 * (1/10**exponentAtPriceOne)
+	geometricExponentIncrementDistanceInTicks := sdkNineDec.Mul(powTen(exponentAtPriceOne.Neg()))
+
+	// Initialize the total price to 1, the current precision to exponentAtPriceOne, and the number of ticks passed to 0
+	totalPrice := sdk.OneDec()
+	ticksPassed := sdk.ZeroInt()
+	exponentAtCurrentTick := exponentAtPriceOne
+
+	// Set the currentAdditiveIncrementInTicks to the exponentAtPriceOne
+	currentAdditiveIncrementInTicks := powTenBigDec(exponentAtPriceOne)
+
+	// Now, we loop through the k increments until we have passed the price
+	// Once we pass the price, we can determine what which geometric exponents we have filled in their entirety,
+	// as well as how many ticks that corresponds to
+	// In the opposite direction (price < 1), we do the same thing (just decrement the geometric exponent instead of incrementing).
+	// The only difference is we must reduce the increment distance by a factor of 10.
+	if price.GT(sdk.OneDec()) {
```

Comment:
Can we abstract this into a helper? I think it would help with encapsulating the complexity and allow for better testability of the helper. I'm cool with making a separate issue for this suggestion